libretroshare/src/rsserver/p3peers.cc

```diff
@@ -1104,10 +1111,139 @@ bool p3Peers::GetPGPBase64StringAndCheckSum(	const RsPgpId& gpg_id,
 	return true ;
 }
 
+bool p3Peers::getShortInvite(
+        std::string& invite, const RsPeerId& pSslId, bool formatRadix,
+        bool includeFingerprint, bool bareBones )
+{
+	const RsPeerId& sslId(pSslId.isNull() ? getOwnId() : pSslId);
+
+	RsPeerDetails dts;
+	if (!getPeerDetails(sslId, dts)) return false;
+
+	RsUrl inviteUrl;
+	inviteUrl
+	        .setScheme("rs")
```

Comment:
why not use the classical retroshare:// ? It exists already and the GUI will recognize it. Furthermore I dont see the point in handlign urls in libretroshare. There's nothing to click in libretroshare. So you should probably keep the invite as a base64 string and let the GUI or any interface make a link out of it. Otherwise you'll have to parse html links  inside libretroshare (and without Qt!!) which is a pain to do securely.

Reply:
No need to parse HTML, we have RsUrl class that does it safely. URI, and HTML are two completely different formats. rs:// and not retroshare:// is for two motivations, it make the invite shorter, and most libraries that are used in lot of places to detect URL and make them clickable on applications around, recognize things like VERY_SHORT_STRING://something quite commonly, while they usually don't for things like LONGER_STRING://something. This way if someone send an RS invite trough whatsapp or telegram or mail it is much more likely that it get clickable and the used do not need to copy paste, which on mobile phones range from a pain in the ass to impossible. The point to handle it in the library is that all applications that uses libretroshare will have the same format, while if we let them handle each of them by them self is duplicated code and incompatibility (like it happened with Qml app as the URL library there behave different then QUrl I had to sweat a lot to make them interoperate). Anyway we should plan to register both retroshare:// and rs:// as URI schemes to IANA

Reply:
If you want the invite to be short, you'd rather use the same format than the one we've been using for a while now in libretroshare: field id + size (in 1,2,5 format) + data, and parse it the same way we parse certificates. It has several advantages over this url thing:
- it is more compact (no something=, just an id, so plenty of space gained, not to mention the rs://)
- no url structure that is totally awkward inside libretroshare and needs some careful parsing, so it's much more robust.
- no parsing ambiguity problems due to encoding because it's all binary. What happens if you have a & or a space in the name of the invite? You'll need to encode this using %foo objects, etc. What about utf8 caracters etc? This is going to create additional weaknesses for no additional value.

Reply:
Moreover, I dont see a single usage case where you would like to create the invite yourself: the invite is created by libretroshare and parsed by libretroshare, which means it doesn't need to have a "readable" url format (with all the risks that come with it). For instance in a webUI you'd call JSonAPI->sendMeAnInvite, you get the base64 string and paste it as needed. Same when adding a friend. If needed,you may implement a API function to parse it for you, such as returning the name, SSL Id, etc for display. That's what we do with certificates.

Reply:
The point to make it look like a normal URL is not programmer comfort to generate it outside libretroshare (on the contrary I want to avoid that!! Right now retroshare URLs are generated in retroshare-gui that created compatibility problems between Qml app and retroshare-gui!!!!), it is for user comfort. To make other applications, like a web browsers, whatsapp, telegram, whatever that may be used to give/get the invite to/from other that you cannot yer reach over RS, recognize the invites as something that might be clicked, because copy paste an invite in mobile phones is really bad and even impossible in some cases, such as when you use whatsapp (that is the app 99% of user uses today). Frankly I don't think that prepending `rs://sslid/` in front of whatever format we agree to use (that could be the plain old radix) make it insecure... I think you should take a look at RsUrl code... it is simple (because it doesn't intend to support every URL in the word but only simple URLs as the one we uses) and safe.

Reply:
What's insecure is to use the url format to store multiple fields some of which are strings. This is the open door to parsing problems and attack vectors. Remember what happenned with xml ;-)

Now what you're describing is exactly my point: to allow all applications to understand the invite in the best level of compatibility, you need to set it to be a base64 string that is unbreakable, and let libretroshare parse it. Having urls in libretroshare itself (especially if it has a single field that is the base64 string itself) doesnt make sense to me.
Then if you want to display this in a way or another in your own app, fine. At least you wont have to parse it in order to extract the actual information you need to re-form your own url format.

Reply:
For the record, retroshare certificate have always been a pain to parse and caused many problems and errors in different applications. Now that we have a certificate format that is a single base64 string, all these problems are gone. It's more robust, more secure and easier to handle. Nothing prevents you in your own app to display the cert a rs://certificate/[base64 string]

Reply:
The problem is not in an App that we control of course there we can do whatever we want and not use URL at all eventually. The problem is in the apps we don't control such as whatsapp, that is what most users could use in mobile phones so exchange the invites, so we can avoid dangerous fields in the URL, but it is important the string is passed around looks like an URL and can be parsed as a simple URL, so those apps "make it blue" and clickable, that automatically trigger an OS handler that look for the application able to handle that URI scheme and pass it to the correct app for handling.

Reply:
Of course. But then how do you get the url to arrive in a whatsapp application? It needs to come from soem UI of some sort. It cannot come directly from libretroshare. So that's where you add the rs://thing

Reply:
> For the record, retroshare certificate in plain PGP key format and named fields have always been a pain to parse and caused many problems and errors in different applications. Now that we have a certificate format that is a single base64 string, all these problems are gone. It's more robust, more secure and easier to handle. Nothing prevents you in your own app to display the cert a rs://certificate/[base64 string]

The point in doing it in libretroshare, is that hen all apps developed on top of libretroshare use the same compatible format, so a link generated by customAndroidRsApp is exactly the same by one generated by retroshare-gui avoiding fragmentation

Reply:
sure. And that's pretty easy to do. You dont need to enforce it, given the cost of it. A simple comment around the method that give you the invite to suggest how to convert it into a url would do the job.

Reply:
> Of course. But then how do you get the url to arrive in a whatsapp application? It needs to come from soem UI of some sort. It cannot come directly from libretroshare. So that's where you add the rs://thing

That means that then one should implement it in the same way in retroshare-gui, elrepo.io, retroshare-qml-app, emoty... that all uses different libraries and are maintained by different persons... and it was already a pain to make it compatible in the past that was only between retroshare-gui and retroshare-qml-app...

Reply:
> sure. And that's pretty easy to do. You dont need to enforce it, given the cost of it. A simple comment around the method that give you the invite to suggest how to convert it into a url would do the job.

It is not enforced at all, the `formatRadix` option is there just for that... I don't see which is the cost to do that in the library... XML and RsUrl two completely different beasts, a simple URL won't introduce any vulnerability